### PR TITLE
[VectorDistribution] Add distribution pattern for vector::ContractionOp

### PR DESCRIPTION
**Pull request by bangtianliu**
Depends on #18822 

This patch adds support for distributing vector::ContractionOp when contraction op is generated from reduction and mfma instruction is not used.

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp [hunk 1]
```diff
@@ -563,6 +563,223 @@ struct DistributeMultiReduction final
...
 };
 
+/// The lowering for Contract is performed in three steps (similar to above
+/// multi_reduction):
+///   1. Local Contract: Each thread performs operations on its locally
+///   distributed elements.
+///   2. Thread Reduction: Threads collectively reduce the results from step 1
+///   across threads,
+///      using a butterfly shuffle if distribution occurs along the reduction
+///      dimension.
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp]: Butterfly shuffle is an implementation detail that's not observable from this pattern. I'd just say we perform subgroup reduction.

Also nit: the formatting is weird.

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp [hunk 2]
```diff
@@ -563,6 +563,223 @@ struct DistributeMultiReduction final
   int64_t maxBitsPerShuffle;
 };
 
+/// The lowering for Contract is performed in three steps (similar to above
+/// multi_reduction):
+///   1. Local Contract: Each thread performs operations on its locally
+///   distributed elements.
+///   2. Thread Reduction: Threads collectively reduce the results from step 1
+///   across threads,
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp]: I'd call it something like: subgroup reduction or inter-thread reduction.

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp [hunk 3]
```diff
@@ -563,6 +563,223 @@ struct DistributeMultiReduction final
...
+///      using a butterfly shuffle if distribution occurs along the reduction
+///      dimension.
+///   3. Accumulator Reduction: Each thread combines its intermediate results
+///   with its held accumulator.
+///
+/// Currently, reduction across multiple warps is not supported.
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp:576]: let's try to stick to the portable naming like in the gpu dialect

```suggestion
/// Currently, reduction across multiple subgroups is not supported.
```

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp [hunk 4]
```diff
@@ -563,6 +563,223 @@ struct DistributeMultiReduction final
...
+/// TODO: Add support for reductions across multiple warps.
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp]: Can you describe why we'd need it?

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp [hunk 5]
```diff
@@ -563,6 +563,223 @@ struct DistributeMultiReduction final
...
+                                PatternRewriter &rewriter) const override {
+    FailureOr<VectorContractOpInfo> maybeOpInfo =
+        VectorContractOpInfo::inferFromIndexingMaps(
+            contractOp.getIndexingMapsArray());
+    if (failed(maybeOpInfo)) {
+      return rewriter.notifyMatchFailure(contractOp, "not a contraction");
+    }
+    // If mmaAttr exists, defer the lowering to use MMA.
+    // Notify failure if the "iree.amdgpu.mma" intrinsic attribute is present.
+    IREE::GPU::MMAAttr mmaAttr =
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp]:
```suggestion
    auto mmaAttr =
```

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp [hunk 6]
```diff
@@ -563,6 +563,223 @@ struct DistributeMultiReduction final
...
+        contractOp->getAttrOfType<IREE::GPU::MMAAttr>("iree.amdgpu.mma");
+    if (mmaAttr) {
+      return rewriter.notifyMatchFailure(
+          contractOp, "iree.amdgpu.mma intrinsic attribute exists");
+    }
+
+    NestedLayoutAttr lhsLayout =
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp]:
```suggestion
    auto lhsLayout =
```

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp [hunk 7]
```diff
@@ -563,6 +563,223 @@ struct DistributeMultiReduction final
...
+        dyn_cast<NestedLayoutAttr>(signature[contractOp.getLhs()]);
+    if (!lhsLayout) {
+      return rewriter.notifyMatchFailure(
+          contractOp, "missing nested layout for contraction lhs");
+    }
+    NestedLayoutAttr rhsLayout =
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp]:
```suggestion
    auto rhsLayout =
```

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp [hunk 8]
```diff
@@ -563,6 +563,223 @@ struct DistributeMultiReduction final
...
+          contractOp, "missing nested layout for contraction rhs");
+    }
+
+    Value disLhs = getDistributed(rewriter, contractOp.getLhs(), lhsLayout);
+    Value disRhs = getDistributed(rewriter, contractOp.getRhs(), rhsLayout);
+
+    Value acc = contractOp.getAcc();
+    Value res = contractOp.getResult();
+    VectorValue accVector = dyn_cast<VectorValue>(acc);
+    VectorValue resVector = dyn_cast<VectorValue>(res);
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp]:
```suggestion
    auto accVector = dyn_cast<VectorValue>(acc);
    auto resVector = dyn_cast<VectorValue>(res);
```

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp [hunk 9]
```diff
@@ -563,6 +563,223 @@ struct DistributeMultiReduction final
...
+      disAcc = getDistributed(rewriter, accVector, signature[accVector]);
+    } else {
+      disAcc = contractOp.getAcc();
+    }
+
+    Location loc = contractOp.getLoc();
+    int64_t rank = lhsLayout.getRank();
+
+    SmallVector<bool> reducedDims(rank, false);
+    auto maps = contractOp.getIndexingMapsArray();
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp]: Use the actual type since it's not obvious based on the RHS only

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp [hunk 10]
```diff
@@ -563,6 +563,223 @@ struct DistributeMultiReduction final
...
+
+    // Identify the reduction dimension and apply it for cross-thread reduction.
+    MLIRContext *ctx = maps[0].getContext();
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp]: I'd put it just before loc and extract from `contractOp` instead.

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp [hunk 11]
```diff
@@ -563,6 +563,223 @@ struct DistributeMultiReduction final
...
+
+    ArrayRef<Attribute> iteratorTypes =
+        contractOp.getIteratorTypes().getValue();
+    SmallVector<Attribute> newIterators;
+
+    // Given that the distribution format is <BATCH x OUTER x ELEMENT>,
+    // the iterations and affine maps need to be replicated three times.
+
+    // Replicate the iterators for local vector.contract
+    for (int i = 0; i < 3; i++) {
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp]: https://llvm.org/docs/CodingStandards.html#prefer-preincrement

```suggestion
    for (int i = 0; i < 3; ++i) {
```

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp [hunk 12]
```diff
@@ -563,6 +563,223 @@ struct DistributeMultiReduction final
...
+      newIterators.append(iteratorTypes.begin(), iteratorTypes.end());
+    }
+
+    // Replicate the affine maps for local vector.contract
+    SmallVector<AffineMap> newMaps;
+    for (auto map : maps) {
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp]: Use the actual type since it's not obvious based on the RHS only

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp [hunk 13]
```diff
@@ -563,6 +563,223 @@ struct DistributeMultiReduction final
...
+      int64_t numDims = map.getNumDims();
+      int64_t numResults = map.getNumResults();
+      SmallVector<AffineExpr> exprs;
+      for (int i = 0; i < 3; i++) {
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp]: also here

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp [hunk 14]
```diff
@@ -563,6 +563,223 @@ struct DistributeMultiReduction final
...
+        AffineMap shiftedMap = map.shiftDims(i * numDims);
+        for (int j = 0; j < numResults; j++) {
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp]: also here

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp [hunk 15]
```diff
@@ -563,6 +563,223 @@ struct DistributeMultiReduction final
...
+
+    if (resVector) {
+      replaceOpWithDistributedValues(rewriter, contractOp, accReduced);
+    } else {
+      Value accReducedVal = rewriter.create<vector::ExtractOp>(
+          loc, accReduction, SmallVector<int64_t>{0});
+      replaceOpWithDistributedValues(rewriter, contractOp, accReducedVal);
+    }
+
+    return success();
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp:701]: Would it be possible to split this function into a couple smaller helpers matching the 3-stage lowering described in the comment? This implementation is a bit long and hard to follow.

File: compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUConfigureTensorLayouts.cpp [hunk 16]
```diff
@@ -341,8 +341,10 @@ struct LLVMGPUConfigureTensorLayoutsPass final
     llvm::StringLiteral scheduleAttrName =
         IREE::GPU::MMAScheduleAttr::getMnemonic();
     DictionaryAttr configDict = getTranslationInfo(func).getConfiguration();
-    auto scheduleAttr = dyn_cast_or_null<IREE::GPU::MMAScheduleAttr>(
-        configDict.get(scheduleAttrName));
+    IREE::GPU::MMAScheduleAttr scheduleAttr = nullptr;
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUConfigureTensorLayouts.cpp]:
```suggestion
    IREE::GPU::MMAScheduleAttr scheduleAttr;
```

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp (hunk 4 above)
Comment (bangtianliu) [reply compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp]: Now the implementation is limited to support only 1 subgroup, I don't know whether 1 subgroup for sure is always the optimal config for performance. if multiple subgroups can achieve more performance benefits, then adding support for reductions across multiple subgroups will be a TODO.

**Comment by bangtianliu** [comment]
All the above comments have been applied to #18800, which includes the content from this patch and has already been merged.  This PR can now be closed.

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp (hunk 4 above)
Comment (Groverkss) [reply compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp]: I think I added this comment for multi_reduction and it was copied over. Since we can describe reduction on subgroups on layouts, if we ever need workgroup reduce with atomic read/write we can do it. Can also just remove the comment if we dont plan on ever implementing that.

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp (hunk 2 above)
Comment (Groverkss) [reply compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp]:
```
Subgroup Reduction: Threads in each subgroup reduce the results from step 1 across threads
```

could work fine

File: compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUConfigureTensorLayouts.cpp [hunk 17]
```diff
@@ -341,8 +341,10 @@ struct LLVMGPUConfigureTensorLayoutsPass final
...
+    if (configDict)
+      scheduleAttr = dyn_cast_or_null<IREE::GPU::MMAScheduleAttr>(
+          configDict.get(scheduleAttrName));
```
Comment (Groverkss) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUConfigureTensorLayouts.cpp]: I think you can always assume there will be a configDict. What you cannot assume is there always being a scheduleAttr. The pass should return error if there is no translation_info / configuration.

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp (hunk 4 above)
Comment (bangtianliu) [reply compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp]: Yes, I can remove it.

File: compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUConfigureVectorLayouts.cpp [hunk 18]
```diff
@@ -268,8 +268,10 @@ struct LLVMGPUConfigureVectorLayoutsPass final
         func->getAttrOfType<IREE::GPU::MMAScheduleAttr>(scheduleAttrName);
     if (!scheduleAttr) {
       DictionaryAttr configDict = getTranslationInfo(func).getConfiguration();
-      scheduleAttr = dyn_cast_or_null<IREE::GPU::MMAScheduleAttr>(
-          configDict.get(scheduleAttrName));
+      if (configDict) {
+        scheduleAttr = dyn_cast_or_null<IREE::GPU::MMAScheduleAttr>(
+            configDict.get(scheduleAttrName));
+      }
     }
```
Comment (Groverkss) [inline compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUConfigureVectorLayouts.cpp]: I think this file got deleted actually. You might want to rebase.

File: compiler/src/iree/compiler/Codegen/Common/GPU/test/gpu_nested_layout_vector_distribution.mlir [hunk 19]
```diff
@@ -1139,3 +1139,118 @@ builtin.module attributes { transform.with_named_sequence } {
...
+  %c128 = arith.constant 128 : index
+  %cst = arith.constant dense<0.000000e+00> : vector<f32>
+  %cst_0 = arith.constant 0.0 : f16
+  %out = scf.for %i = %c0 to %c128 step %c1 iter_args(%arg0 = %cst) -> (vector<f32>) {
+    %root = vector.transfer_read %arr[%c0, %c0], %cst_0 {in_bounds = [true, true]} : memref<32x32xf16>, vector<32x32xf16>
+    %rootl = iree_vector_ext.to_layout %root to layout(#layout) : vector<32x32xf16>
+    %b = arith.addf %rootl, %a : vector<32x32xf16>
+    %c = arith.extf %b : vector<32x32xf16> to vector<32x32xf32>
+    %init = vector.extractelement %arg0[] : vector<f32>
+    %root_red = vector.contract #contraction_trait %c, %c,  %init : vector<32x32xf32>, vector<32x32xf32> into f32
```
Comment (Groverkss) [inline compiler/src/iree/compiler/Codegen/Common/GPU/test/gpu_nested_layout_vector_distribution.mlir:1228]: Can we have a test where vector.contract has a parallel dimension as well?

**Review by Groverkss — APPROVED** [review]
